typescript/packages/jumble/src/contexts/CharmManagerContext.tsx

```diff
@@ -0,0 +1,39 @@
+import React, { createContext, useContext, useEffect, useRef } from "react";
+import { CharmManager } from "@commontools/charm";
+import { replica } from "@/views/state";
+import { effect } from "@commontools/runner";
+
+export type CharmManagerContextType = {
+  charmManager: CharmManager;
+};
+
+const CharmManagerContext = createContext<CharmManagerContextType>(null!);
+
+const defaultManager = new CharmManager(undefined, "memory");
+
+export const CharmsManagerProvider: React.FC<{ children: React.ReactNode }> = ({ children }) => {
+  const [charmManager, setCharmManager] = React.useState<CharmManager>(defaultManager);
+
+  useEffect(() => {
+    const cleanup = effect(replica, (currentReplica) => {
+      // Create new charm manager instance with updated replica
+      const storageType = (import.meta as any).env.VITE_STORAGE_TYPE ?? "remote";
+      const manager = new CharmManager(currentReplica, storageType);
+      manager.init();
+      console.log("Started CharmManager for replica " + currentReplica, manager);
+      setCharmManager(manager);
+    });
+
+    return () => {
+      cleanup();
+    };
+  }, [setCharmManager]);
```

Comment:
This will teardown and spin up the CharmManager instance if the replica is changed, but I'm not sure it's working 100% of the time to switch replicas at runtime still.